File: mpl/mpli.c

#include <stdlib.h>
#include <stdio.h>

typedef char *mplword;

/* #define WORDSIZE (sizeof(mplword)) */
#define WORDSIZE 8

FILE *script;
FILE *input;
FILE *output;

#define MEMSIZE 0x3000
union
{
 char c[WORDSIZE*MEMSIZE];
 mplword w[MEMSIZE];
} mem;

int adr;
char c;
int goon;
int trace;
char instr;
long pc, r0, r1, rc, ru;
long t;
int state;
#define RUN 0
#define CODE 1
char buf[1000];

openfile (char *mode)
{
 char filename[1000];
 int i;
 /* printf ("openfile r1=%X mem[r1]=%X\n", r1, mem.c[r1]); */
 i = 0;
 while (mem.c[r1+i])
 {
  filename[i] = mem.c[r1+i];
  i++;
 }
 filename[i] = 0;
 /* printf ("open file (%s) with mode (%s)\n", filename, mode); */
 r1 = (long) fopen (filename, mode);
}

main (int argc, char **argv)
{
	/* printf ("WORDSIZE=%X memory size=%X sizeof(char)=%X\n", WORDSIZE, sizeof(mem), sizeof(char)); */
	script = fopen (argv[1], "r");
	if (script == NULL)
	{
		perror (argv[1]);
		exit(0);
	}
	if ((argc < 3) || !strcmp (argv[2], "-"))
		input = stdin;
	else
	{
		input = fopen (argv[2], "r");
		if (input == NULL)
		{
			perror (argv[2]);
			exit(0);
		}
	}
	if ((argc < 4) || !strcmp (argv[3], "-"))
		output = stdout;
	else
	{
		output = fopen (argv[3], "w");
		if (output == NULL)
		{
			perror (argv[3]);
			exit(0);
		}
	}

	adr = 0;
	while (!feof(script))
	{
		c = fgetc (script);
		mem.c[adr++] = c;
	}

	trace = 0;
	goon = 1;
	pc = 0;
	r0 = 0;
	r1 = 0;
	rc = adr;

	/* printf ("Running...\n"); */
	while (goon)
	{
		/* printf (" pc=%X ", pc); */
		instr = mem.c[pc++];
		/* printf (" ok1 "); */
		if (trace & 1)
		{ 
		 /* printf (" ok3 "); */
		 /* printf (" instr=%X ", instr); */
		 putchar(instr);
		 /* printf (" ok4 "); */
		}
		/* printf (" ok2 "); */
		if (trace & 2)
		{
			printf ("\nR0=%4lX\tR1=%4lX\tRC=%4lX\tPC=%4lX\tinstr=%c ",
				r0, r1, rc, pc, instr);
			gets(buf);
		}
		fflush (stdout);
		if (state == CODE)
		{
			if (instr == '}')
				state = RUN;
			else
				mem.c[rc++] = instr;
		}
		else switch (instr)
		{
			case ',': r1 = r0; break;
			case 'X': t = r0; r0 = r1; r1 = t; break;
			case 'p': r0 = pc; break;
			case 'Q': goon = 0; break;
			case '?': if (r1 == 0) { r1 = pc; pc = r0; } break;
			/*
			case 'R': r0 = (long)(mem.w[r0/WORDSIZE]); break;
			case 'W': mem.w[r0/WORDSIZE] = (mplword)r1; break;
			*/
			case 'R': r0 = *(long *)(mem.c+r0); break;
			case 'W': if ((r0 < 0) || (r0 >= sizeof(mem))) { printf ("W : Write out of memory\n"); exit(0); } 
				  *(long *)(mem.c+r0) = r1; break;
			case 'r': r0 = mem.c[r0]; break;
			case 'w': if ((r0 < 0) || (r0 >= sizeof(mem))) { printf ("w : Write out of memory\n"); exit(0); }
				  mem.c[r0] = r1; break;
			case '#': r0 = 0; break;

			case '0': r0 = r0*16; break;
			case '1': r0 = r0*16+1; break;
			case '2': r0 = r0*16+2; break;
			case '3': r0 = r0*16+3; break;
			case '4': r0 = r0*16+4; break;
			case '5': r0 = r0*16+5; break;
			case '6': r0 = r0*16+6; break;
			case '7': r0 = r0*16+7; break;
			case '8': r0 = r0*16+8; break;
			case '9': r0 = r0*16+9; break;
			case 'A': r0 = r0*16+10; break;
			case 'B': r0 = r0*16+11; break;
			case 'C': r0 = r0*16+12; break;
			case 'D': r0 = r0*16+13; break;
			case 'E': r0 = r0*16+14; break;
			case 'F': r0 = r0*16+15; break;

			case '~': r0 = ~r0; break;
			case '+': r0 = r0 + r1; break;
			case '-': r0 = r0 - r1; break;
			case '*': r0 = r0 * r1; break;
			case '/': r0 = r0 / r1; break;
			case '%': r0 = r0 % r1; break;
			case '&': r0 = r0 & r1; break;
			case '^': r0 = r0 ^ r1; break;
			case '|': r0 = r0 | r1; break;
			case '<': r0 = r0 << r1; break;
			case '>': r0 = r0 >> r1; break;

			case 'G': if (feof(input)) r0 = -1; else r0 = fgetc(input); break;
			case 'P': fputc (r0, output); break;
			case 'H': rc = r0; break;
			case 'h': r0 = rc; break;
			case '{': state = CODE; break;
			case '}': if ((rc < 0) || (rc >= sizeof(mem))) { printf ("} : Write out of memory\n"); exit(0); }
				  mem.c[rc++] = instr; break;
			case '_': if ((rc < 0) || (rc >= sizeof(mem))) { printf ("_ : Write out of memory\n"); exit(0); }
				  mem.c[rc++] = r0; break;
			case 's': r0 = WORDSIZE; break;

			case 'S':
			 switch (r0)
			 {
			  case 0x11: openfile ("rb"); break;
			  case 0x12: openfile ("wb"); break;
			  case 0x13: openfile ("ab"); break;
			  case 0x14: openfile ("r+b"); break;
			  case 0x15: openfile ("w+b"); break;
			/*
			  case 0x1F: fclose ((FILE *)(mem.w[r1/WORDSIZE])); break;
			  case 0x20: mem.w[(r1+WORDSIZE)/WORDSIZE] = (mplword) fgetc ((FILE *)(mem.w[r1/WORDSIZE])); break;
			  case 0x21: fputc ((char)(mem.w[(r1+WORDSIZE)/WORDSIZE]), (FILE *)(mem.w[r1/WORDSIZE])); break;
			  case 0x1F: fclose (*(FILE **)(mem.c+r1)); break;
			*/
			  case 0x1F: fclose ((FILE *) r1); break;
			  case 0x20: *(long *)(mem.c+r1+WORDSIZE) = fgetc (*(FILE **)(mem.c+r1)); break;
			  case 0x21 : fputc (*(long *)(mem.c+r1+WORDSIZE), *(FILE **)(mem.c+r1)); break;
			 }

			 break;

			case 'T': trace = 2; break;
			case 't': trace = 1; break;
			case 'e': printf ("\nError\n"); goon = 0; break;

			case 'u': r0 = ru; break;
			case 'U': ru = r0; break;
			case 'J': t = r0;
				  r0 = *(long *)(mem.c+t);
				  r1 = *(long *)(mem.c+t+WORDSIZE);
				  pc = *(long *)(mem.c+t+2*WORDSIZE);
				  break;
			case ' ':
			case '\n':
			case '\r':
			case '\t':
			case ';': break;

			default: 
			 *(long *)(mem.c+ru) = r0;
			 *(long *)(mem.c+ru+WORDSIZE) = r1;
			 *(long *)(mem.c+ru+2*WORDSIZE) = pc;
			 pc = ru+3*WORDSIZE;
			 break;
		}
	}
}

File: mpl/mpli.pl
#!/usr/bin/perl

my $wordsize = 8;

my $trace = 0;

my $nargs = @ARGV;
my $script;
my $inputf;
my $outputf;
my $input;
my $output;

if ($nargs >= 1)
{ 
 $script = $ARGV[0]; 
 $inputf = $ARGV[1];
 $outputf = $ARGV[2];
} 
else
{
 print "Script ? ";
 $script = <STDIN>;
 chomp $script;

 print "Input file ? ";
 $inputf = <STDIN>;
 chomp $inputf;
 
 print "Output file ? ";
 $outputf = <STDIN>;
 chomp $outputf;
}

open SCRIPT, $script or die "$script: $!\n";

if (($inputf eq '') or ($inputf eq '-'))
{
 $input = *STDIN;
}
else
{
 open INPUT, $inputf or die "$inputf: $!\n";
 $input = *INPUT;
}

if (($outputf eq '') or ($outputf eq '-'))
{
 $output = *STDOUT;
}
else
{
 open OUTPUT, ">$outputf" or die "$outputf: $!\n";
 $output = *OUTPUT;
}
 

my @mem = ();
my $adr = 0;
my $state = 'normal';

# print "Reading code...\n";
while (<SCRIPT>)
{
 if ($trace != 0) { print $_; }
 @chars = split //, $_;
 foreach $char (@chars)
 {
  $mem[$adr++] = ord $char;    
 }
}
# print (sprintf "program size: %X\n", $adr);

my $goon = 1;
my $pc = 0;
my $r0 = 0;
my $r1 = 0;
my $rc = $adr;
my $ru = 0;
my $t;
my $instr;
$state = 'normal';
my $fh;

sub openfile 
{
 my $mode = shift;
 my $name = "";
 my $i = 0;
 # print "\nr1=" . (sprintf "%X",$r1) . " mem[r1]=" . (sprintf "%X", $mem[$r1]) . " i=$i\n";
 while ($mem[$r1+$i]!=0)
 {
  $name = $name . chr($mem[$r1+$i]);
  $i++;
 }
 # print "open file ($name) with mode ($mode)\n";
 open $fh, "$mode$name" or die "$name: $!\n";
 $r1 = $fh;
}

# print "Running...\n";
# $mem[0x7F00] = 0x30;
while ($goon != 0)
{
 # if ($mem[0x7F00] != 0x30) { $goon = 0; }
 # if (($trace == 1) && ($mem[0x3460] != 0x4000)) { $goon = 0; }
 $instr = chr($mem[$pc++]);
 if ($trace & 1) { print $instr; }

 if ($trace & 2) { 
  print ((sprintf "\nr0=%4X\tr1=%4X\trc=%4X\tru=%4X\tpc=%4X\tinstr=%s ",$r0,$r1,$rc,$ru,$pc,$instr) . (ord $instr) . " "); 
  my $r = <STDIN>; 
 }

 if ($state eq 'code') 
 {
  if ($instr eq '}') { $state = 'normal'; }
  else { $mem[$rc++] = ord $instr; }
 }
 elsif ($instr eq ',') { $r1 = $r0; }
 elsif ($instr eq 'X') { $t = $r0; $r0 = $r1; $r1 = $t; }
 elsif ($instr eq 'p') { $r0 = $pc; }
 elsif ($instr eq 'Q') { $goon = 0; }
 elsif ($instr eq '?') { if ($r1 == 0) { $r1 = $pc; $pc = $r0; } }
 elsif ($instr eq 'R') { $r0 = $mem[$r0]; }
 elsif ($instr eq 'W') { $mem[$r0] = $r1; }
 elsif ($instr eq 'r') { $r0 = $mem[$r0]; }
 elsif ($instr eq 'w') { $mem[$r0] = $r1; }
 elsif ($instr eq '#') { $r0 = 0; }

 elsif ($instr eq '0') { $r0 = $r0*16; }
 elsif ($instr eq '1') { $r0 = $r0*16+1; }
 elsif ($instr eq '2') { $r0 = $r0*16+2; }
 elsif ($instr eq '3') { $r0 = $r0*16+3; }
 elsif ($instr eq '4') { $r0 = $r0*16+4; }
 elsif ($instr eq '5') { $r0 = $r0*16+5; }
 elsif ($instr eq '6') { $r0 = $r0*16+6; }
 elsif ($instr eq '7') { $r0 = $r0*16+7; }
 elsif ($instr eq '8') { $r0 = $r0*16+8; }
 elsif ($instr eq '9') { $r0 = $r0*16+9; }
 elsif ($instr eq 'A') { $r0 = $r0*16+10; }
 elsif ($instr eq 'B') { $r0 = $r0*16+11; }
 elsif ($instr eq 'C') { $r0 = $r0*16+12; }
 elsif ($instr eq 'D') { $r0 = $r0*16+13; }
 elsif ($instr eq 'E') { $r0 = $r0*16+14; }
 elsif ($instr eq 'F') { $r0 = $r0*16+15; }

 elsif ($instr eq '~') { $r0 = -1 ^ $r0; }
 elsif ($instr eq '+') { $r0 = $r0 + $r1; }
 elsif ($instr eq '-') { $r0 = $r0 - $r1; }
 elsif ($instr eq '*') { $r0 = $r0 * $r1; }
 elsif ($instr eq '/') { $r0 = int ($r0 / $r1); }
 elsif ($instr eq '%') { $r0 = $r0 % $r1; }
 elsif ($instr eq '&') { $r0 = $r0 & $r1; }
 elsif ($instr eq '^') { $r0 = $r0 ^ $r1; }
 elsif ($instr eq '|') { $r0 = $r0 | $r1; }
 elsif ($instr eq '<') { $r0 = $r0 << $r1; }
 elsif ($instr eq '>') { $r0 = $r0 >> $r1; }

 elsif ($instr eq 'G') { if (eof($input)) { $r0 = -1; } else { $r0 = ord getc($input); } }
 elsif ($instr eq 'P') { print $output chr($r0); }
 elsif ($instr eq 'H') { $rc = $r0; }
 elsif ($instr eq 'h') { $r0 = $rc; }
 elsif ($instr eq '{') { $state = 'code'; }
 elsif ($instr eq '}') { $mem[$rc++] = ord $instr; }
 elsif ($instr eq '_') { $mem[$rc++] = $r0; }
 elsif ($instr eq 's') { $r0 = $wordsize; }

 elsif ($instr eq 'S')
 {
  if ($r0 == 0x11) { openfile "<"; }
  elsif ($r0 == 0x12) { openfile ">"; }
  elsif ($r0 == 0x13) { openfile ">>"; }
  elsif ($r0 == 0x14) { openfile "+<"; }
  elsif ($r0 == 0x15) { openfile "+>"; }
  # elsif ($r0 == 0x1F) { close $r1; }
  elsif ($r0 == 0x1F) { close $mem[$r1]; }
  # elsif ($r0 == 0x20) { $r0 = ord getc($r1); }
  elsif ($r0 == 0x20) { $mem[$r1+$wordsize] = ord getc($mem[$r1]); }
  elsif ($r0 == 0x21) { $fh = $mem[$r1]; print $fh chr($mem[$r1+$wordsize]); }
 }

 elsif ($instr eq 'T') { $trace = 2; }
 elsif ($instr eq 't') { $trace=1; } 
 elsif ($instr eq 'e') { print "\nError\n"; $goon = 0; }

 elsif ($instr eq 'u') { $r0 = $ru; }
 elsif ($instr eq 'U') { $ru = $r0; }
 elsif ($instr eq 'J') { $t = $r0; 
                         $r0 = $mem[$t];
                         $r1 = $mem[$t+$wordsize];
                         $pc = $mem[$t+2*$wordsize]; }
 elsif (($instr eq ' ') || 
        ($instr eq "\n") ||
        ($instr eq "\r") ||
        ($instr eq "\t") ||
        ($instr eq ";")) { }
 else {
  # print (sprintf "undefined pc=%X ru=%X\n", $pc, $ru);
  $mem[$ru] = $r0;
  $mem[$ru+$wordsize] = $r1;
  $mem[$ru+2*$wordsize] = $pc;
  $pc = $ru+3*$wordsize;
  # print (sprintf "undefined pc=%X ru=%X\n", $pc, $ru);
 }

}

# print "\nDone.\n";

close SCRIPT;
close INPUT;
close OUTPUT;



